### PR TITLE
fix(animations): stringify state values before running the animation

### LINKED ISSUE #12337: Animations: transition doesn't support true/false values when the trigger is a boolean
**I'm submitting a ...**  (check one with "x")

```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When using animation on a component with a boolean trigger, setting a transition from 'false => true' won't do the transition. If you set it to '0 => 1' the transition will work fine.

**Expected behavior**
When using animation on a component with a boolean trigger, setting a transition from 'false => true' should do the specified transition.

**Minimal reproduction of the problem with instructions**
http://plnkr.co/edit/dVVesfq6TCiOclwxRSgf

**What is the motivation / use case for changing the behavior?**
Since state supports 'true' and 'false' values, for the sake of consistency transition should support them as well.

**Please tell us about your environment:**
macOS Sierra, NPM
- **Angular version:** 2.0.1
- **Browser:** all
- **Language:** TypeScript 2.0.3
- **Node (for AoT issues):** `node --version` =   v6.3.0

### PR DESCRIPTION
**Pull request by matsko**
Closes #12337

### CODE REVIEW

**Comment by vicb** [comment]
status ?

File: modules/@angular/compiler/src/compiler_util/render_util.ts [hunk 1]
```diff
@@ -154,3 +155,7 @@ export function createCheckAnimationBindingStmts(
   const checkDetachStmts: o.Statement[] = [...evalResult.stmts, ...detachStmts];
   return {checkUpdateStmts, checkDetachStmts};
 }
+
+function strigify(exp: o.Expression): o.Expression {
+  return exp.equals(o.NULL_EXPR).conditional(o.literal('null'), exp.callMethod('toString', []));
```
Comment (mhevery) [inline modules/@angular/compiler/src/compiler_util/render_util.ts]: I think `'' + exp` is more efficient, then `exp == null ? 'null' : exp.toString()`

File: modules/@angular/compiler/src/compiler_util/render_util.ts [hunk 2]
```diff
@@ -99,19 +99,20 @@ export function createCheckAnimationBindingStmts(
...
-      animationTransitionVar
-          .set(animationFnExpr.callFn([
-            view, renderElement, isFirstViewCheck(view).conditional(emptyStateValue, oldValue),
-            evalResult.currValExpr
-          ]))
-          .toDeclStmt());
+  updateStmts.push(animationTransitionVar
+                       .set(animationFnExpr.callFn([
+                         view, renderElement,
+                         isFirstViewCheck(view).conditional(emptyStateValue, strigify(oldValue)),
```
Comment (mhevery) [inline modules/@angular/compiler/src/compiler_util/render_util.ts]: I think this change would push `string` event when the transition is a function. Would it not be better if we assumed that all transition are functions, and just have a factory such as `transitionFactory('foo => bar')` That way the `transitionFactory` can handle the stringification internally but not pollute the code. (catch me on VC if you want more detailed explanation.)

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 3]
```diff
@@ -112,6 +112,65 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+           TestBed.overrideComponent(DummyIfCmp, {
+             set: {
+               template: `
+                <div *ngIf="exp2" [@myAnimation]="exp" (@myAnimation.start)="callback($event)"></div>
+              `,
+               animations: [trigger(
+                   'myAnimation',
+                   [
+                     transition('true <=> false', animate(543)),
+                     transition('false => void', animate(100))
```
Comment (mhevery) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: what about 
```
transition((from, to) => true, ...) 
```

I don't think you should assume that `from` and `to` are strings

**Review by mhevery — CHANGES_REQUESTED** [review]

**Comment by willshowell** [comment]
Looking forward to this PR! What's the current status?

**Pull request closed without merging**